src/controllers/media.ts

```diff
@@ -338,3 +338,147 @@ export const getByImdbID = async(ctx: ParameterizedContext<any, Router.IRouterPa
     throw new MediaNotFoundError();
   }
 };
+
+export const getVideo = async(ctx: ParameterizedContext<any, Router.IRouterParamContext<any, {}>>): Promise<any> => {
+  const { title, osdbHash, imdbID } = ctx.query;
+  let { episode, season, year, filebytesize } = ctx.query;
+  [episode, season, year, filebytesize] = [episode, season, year, filebytesize].map(param => param ? Number(param) : null);
+
+  if (!title && !osdbHash && !imdbID) {
+    throw new ValidationError('title, osdbHash or imdbId is a required parameter');
+  }
+
+  if (osdbHash && !filebytesize) {
+    throw new ValidationError('filebytesize is required when passing osdbHash');
+  }
+  let isOnlyOpenSubsSearch = false;
+
+  const query = [];
+  const failedQuery = [];
+
+  if (osdbHash) {
+    query.push({ osdbHash });
+    failedQuery.push({ osdbHash });
+    if (!imdbID && !title) {
+      isOnlyOpenSubsSearch = true;
+    }
+  }
+
+  if (imdbID) {
+    query.push({ imdbID });
+    failedQuery.push({ imdbId: imdbID });
+  }
+
+  if (title) {
+    const titleQuery: any = { searchMatches: { $in: [title] } };
+    const titleFailedQuery: any = { title };
+
+    if (year) {
+      titleQuery.year = year;
+      titleFailedQuery.year = year;
+    }
+    if (episode) {
+      titleQuery.episode = episode;
+      titleFailedQuery.episode = episode;
+    }
+    if (season) {
+      titleQuery.season = season;
+      titleFailedQuery.season = season;
+    }
+    query.push(titleQuery);
+    failedQuery.push(titleFailedQuery);
+  }
+
+  // find an existing metadata record, or previous failure record
+  const [existingResult, existingFailedResult] = await Promise.all([
```

Comment:
Not sure about doing these both at the same time when we may only need one. Can we check them separately to potentially save the double lookup?

Reply:
Hmm, what case would come first then? If we do failed first, actual metadata records are going to be slower to find as it first has to wait for the failed query to finish, or vice versa

Reply:
I'd go failed first since there are always going to be more failed than successes